File: android_native/app/src/main/java/com/example/widgetgallery/presentation/screens/SettingsSheet.kt
package com.example.widgetgallery.presentation.screens

import androidx.compose.animation.core.animateDpAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.BrightnessAuto
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.DarkMode
import androidx.compose.material.icons.filled.LightMode
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.example.widgetgallery.domain.AppLanguage
import com.example.widgetgallery.domain.ColorFlavor
import com.example.widgetgallery.domain.ThemeMode
import com.example.widgetgallery.presentation.AppCredit
import com.example.widgetgallery.presentation.i18n.LocalStrings
import com.example.widgetgallery.presentation.state.AppSettings

@OptIn(ExperimentalMaterial3Api::class, ExperimentalLayoutApi::class)
@Composable
fun SettingsSheet(settings: AppSettings, onDismiss: () -> Unit) {
    val s = LocalStrings.current
    ModalBottomSheet(onDismissRequest = onDismiss) {
        Column(
            Modifier
                .verticalScroll(rememberScrollState())
                .padding(24.dp, 4.dp, 24.dp, 28.dp),
        ) {
            Text(s.appearance, style = MaterialTheme.typography.headlineSmall,
                fontWeight = FontWeight.Bold)
            Spacer(Modifier.height(24.dp))

            Text(s.languageLabel, fontWeight = FontWeight.SemiBold)
            Spacer(Modifier.height(12.dp))
            FlowRow(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                AppLanguage.entries.forEach { lang ->
                    FilterChip(
                        selected = settings.language == lang,
                        onClick = { settings.language = lang },
                        label = { Text(lang.nativeName) },
                    )
                }
            }
            Spacer(Modifier.height(28.dp))

            Text(s.colorLabel, fontWeight = FontWeight.SemiBold)
            Spacer(Modifier.height(14.dp))
            FlowRow(horizontalArrangement = Arrangement.spacedBy(16.dp),
                verticalArrangement = Arrangement.spacedBy(16.dp)) {
                ColorFlavor.entries.forEach { flavor ->
                    Swatch(Color(flavor.seed),
                        settings.flavor == flavor) { settings.flavor = flavor }
                }
            }
            Spacer(Modifier.height(28.dp))

            Text(s.themeLabel, fontWeight = FontWeight.SemiBold)
            Spacer(Modifier.height(12.dp))
            SingleChoiceSegmentedButtonRow(Modifier.fillMaxWidth()) {
                val modes = listOf(
                    Triple(ThemeMode.System, s.themeSystem, Icons.Filled.BrightnessAuto),
                    Triple(ThemeMode.Light, s.themeLight, Icons.Filled.LightMode),
                    Triple(ThemeMode.Dark, s.themeDark, Icons.Filled.DarkMode),
                )
                modes.forEachIndexed { i, (mode, label, icon) ->
                    SegmentedButton(
                        selected = settings.themeMode == mode,
                        onClick = { settings.setThemeMode(mode) },
                        shape = SegmentedButtonDefaults.itemShape(i, modes.size),
                        icon = { Icon(icon, null, Modifier.size(18.dp)) },
                        label = { Text(label) },
                    )
                }
            }
            Spacer(Modifier.height(28.dp))
            Button(onClick = onDismiss, modifier = Modifier.fillMaxWidth()) {
                Text(s.done)
            }
            AppCredit()
        }
    }
}

@Composable
private fun Swatch(color: Color, selected: Boolean, onClick: () -> Unit) {
    val size by animateDpAsState(if (selected) 54.dp else 52.dp, label = "sw")
    val ring = MaterialTheme.colorScheme.onSurface
    Box(
        Modifier.size(size).clip(CircleShape).background(color)
            .border(if (selected) 3.dp else 0.dp,
                if (selected) ring else Color.Transparent, CircleShape)
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        if (selected) Icon(Icons.Filled.Check, null, tint = Color.White)
    }
}

File: lib/presentation/views/settings_sheet.dart
import 'package:flutter/material.dart';

import '../../domain/entities/app_language.dart';
import '../../domain/entities/color_flavor.dart';
import '../../l10n/gen/app_localizations.dart';
import '../state/app_settings_controller.dart';
import '../widgets/app_credit.dart';

/// A polished modal sheet for switching language, accent color and
/// theme mode. Reads / writes [AppSettingsController] via [AppScope].
class SettingsSheet extends StatelessWidget {
  const SettingsSheet({super.key});

  static Future<void> show(BuildContext context) {
    return showModalBottomSheet<void>(
      context: context,
      isScrollControlled: true,
      showDragHandle: true,
      backgroundColor: Theme.of(context).colorScheme.surface,
      shape: const RoundedRectangleBorder(
        borderRadius: BorderRadius.vertical(top: Radius.circular(28)),
      ),
      builder: (_) => const SettingsSheet(),
    );
  }

  @override
  Widget build(BuildContext context) {
    final l = AppLocalizations.of(context);
    final theme = Theme.of(context);
    final settings = AppScope.of(context);

    return SafeArea(
      child: SingleChildScrollView(
        child: Padding(
          padding: const EdgeInsets.fromLTRB(24, 4, 24, 28),
          child: Column(
            mainAxisSize: MainAxisSize.min,
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
            Text(
              l.appearance,
              style: theme.textTheme.headlineSmall
                  ?.copyWith(fontWeight: FontWeight.bold),
            ),
            const SizedBox(height: 24),

            _SectionLabel(icon: Icons.translate, label: l.languageLabel),
            const SizedBox(height: 12),
            Wrap(
              spacing: 10,
              runSpacing: 10,
              children: [
                for (final lang in AppLanguage.values)
                  ChoiceChip(
                    label: Text(lang.nativeName),
                    selected: settings.language == lang,
                    onSelected: (_) => settings.setLanguage(lang),
                  ),
              ],
            ),
            const SizedBox(height: 28),

            _SectionLabel(icon: Icons.palette_outlined, label: l.colorLabel),
            const SizedBox(height: 14),
            Wrap(
              spacing: 16,
              runSpacing: 16,
              children: [
                for (final flavor in ColorFlavor.values)
                  _ColorSwatch(
                    color: Color(flavor.seedValue),
                    selected: settings.flavor == flavor,
                    onTap: () => settings.setFlavor(flavor),
                  ),
              ],
            ),
            const SizedBox(height: 28),

            _SectionLabel(
                icon: Icons.brightness_6_outlined, label: l.themeLabel),
            const SizedBox(height: 12),
            SegmentedButton<ThemeMode>(
              segments: [
                ButtonSegment(
                  value: ThemeMode.system,
                  icon: const Icon(Icons.brightness_auto),
                  label: Text(l.themeSystem),
                ),
                ButtonSegment(
                  value: ThemeMode.light,
                  icon: const Icon(Icons.light_mode_outlined),
                  label: Text(l.themeLight),
                ),
                ButtonSegment(
                  value: ThemeMode.dark,
                  icon: const Icon(Icons.dark_mode_outlined),
                  label: Text(l.themeDark),
                ),
              ],
              selected: {settings.themeMode},
              onSelectionChanged: (s) => settings.setThemeMode(s.first),
            ),
            const SizedBox(height: 28),

            SizedBox(
              width: double.infinity,
              child: FilledButton(
                onPressed: () => Navigator.of(context).pop(),
                style: FilledButton.styleFrom(
                  padding: const EdgeInsets.symmetric(vertical: 16),
                  shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(14)),
                ),
                child: Text(l.done),
              ),
            ),
            const AppCredit(),
          ],
          ),
        ),
      ),
    );
  }
}

class _SectionLabel extends StatelessWidget {
  const _SectionLabel({required this.icon, required this.label});

  final IconData icon;
  final String label;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    return Row(
      children: [
        Icon(icon, size: 18, color: theme.colorScheme.primary),
        const SizedBox(width: 8),
        Text(
          label,
          style: theme.textTheme.titleMedium
              ?.copyWith(fontWeight: FontWeight.w600),
        ),
      ],
    );
  }
}

class _ColorSwatch extends StatelessWidget {
  const _ColorSwatch({
    required this.color,
    required this.selected,
    required this.onTap,
  });

  final Color color;
  final bool selected;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final ring = Theme.of(context).colorScheme.onSurface;
    return GestureDetector(
      onTap: onTap,
      child: AnimatedContainer(
        duration: const Duration(milliseconds: 220),
        curve: Curves.easeOut,
        width: 52,
        height: 52,
        decoration: BoxDecoration(
          color: color,
          shape: BoxShape.circle,
          border: Border.all(
            color: selected ? ring : Colors.transparent,
            width: 3,
          ),
          boxShadow: [
            BoxShadow(
              color: color.withValues(alpha: 0.45),
              blurRadius: selected ? 16 : 6,
              spreadRadius: selected ? 1 : 0,
            ),
          ],
        ),
        child: AnimatedScale(
          scale: selected ? 1 : 0,
          duration: const Duration(milliseconds: 220),
          curve: Curves.easeOutBack,
          child: const Icon(Icons.check, color: Colors.white, size: 24),
        ),
      ),
    );
  }
}
